File: backend/app/schemas/training.py
"""Pydantic schemas for Machine Learning training, cross-validation, and metric evaluation."""

import uuid
from enum import Enum
from typing import Any, Dict, List

from pydantic import BaseModel, Field


class ModelAlgorithm(str, Enum):
    LOGISTIC_REGRESSION = "logistic_regression"
    DECISION_TREE_CLASSIFIER = "decision_tree_classifier"
    RANDOM_FOREST_CLASSIFIER = "random_forest_classifier"
    GRADIENT_BOOSTING_CLASSIFIER = "gradient_boosting_classifier"
    KNN_CLASSIFIER = "knn_classifier"
    LINEAR_REGRESSION = "linear_regression"
    RIDGE_REGRESSION = "ridge_regression"
    LASSO_REGRESSION = "lasso_regression"
    DECISION_TREE_REGRESSOR = "decision_tree_regressor"
    RANDOM_FOREST_REGRESSOR = "random_forest_regressor"
    GRADIENT_BOOSTING_REGRESSOR = "gradient_boosting_regressor"
    KNN_REGRESSOR = "knn_regressor"


class CrossValidationConfig(BaseModel):
    n_splits: int = Field(default=5, ge=2, le=10)
    stratify: bool = True
    shuffle: bool = True
    random_state: int = 42


class TrainingJobCreate(BaseModel):
    project_id: uuid.UUID
    dataset_id: uuid.UUID
    pipeline_id: str | None = None
    target_column: str
    task_type: str  # 'classification' | 'regression'
    algorithm: ModelAlgorithm = ModelAlgorithm.RANDOM_FOREST_CLASSIFIER
    hyperparameters: Dict[str, Any] = Field(default_factory=dict)
    cv_config: CrossValidationConfig | None = Field(default_factory=CrossValidationConfig)
    model_name: str | None = None


class EvaluationMetrics(BaseModel):
    # Classification
    accuracy: float | None = None
    precision: float | None = None
    recall: float | None = None
    f1_score: float | None = None
    roc_auc: float | None = None
    log_loss: float | None = None
    # Regression
    mse: float | None = None
    rmse: float | None = None
    mae: float | None = None
    r2_score: float | None = None
    explained_variance: float | None = None


class ConfusionMatrixData(BaseModel):
    labels: List[str]
    matrix: List[List[int]]
    normalized_matrix: List[List[float]]


class CurvePoint(BaseModel):
    x: float
    y: float
    threshold: float | None = None


class CVFoldResult(BaseModel):
    fold: int
    train_score: float
    val_score: float
    metrics: Dict[str, float]


class CVSummary(BaseModel):
    mean_val_score: float
    std_val_score: float
    folds: List[CVFoldResult]


class FeatureImportanceItem(BaseModel):
    feature: str
    importance: float
    rank: int


class ResidualPoint(BaseModel):
    actual: float
    predicted: float
    residual: float


class TrainingJobDetailResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    dataset_id: uuid.UUID
    model_id: uuid.UUID | None = None
    status: str
    algorithm: str
    task_type: str
    target_column: str
    hyperparameters: Dict[str, Any]
    train_metrics: EvaluationMetrics
    test_metrics: EvaluationMetrics
    cv_summary: CVSummary | None = None
    feature_importances: List[FeatureImportanceItem] = Field(default_factory=list)
    confusion_matrix: ConfusionMatrixData | None = None
    roc_curve: List[CurvePoint] = Field(default_factory=list)
    residuals_sample: List[ResidualPoint] = Field(default_factory=list)
    training_duration_ms: float
    artifact_path: str | None = None
    created_at: str


class AlgorithmInfo(BaseModel):
    algorithm: str
    name: str
    task_type: str
    description: str
    default_hyperparameters: Dict[str, Any]
    hyperparameter_schema: Dict[str, Any]

File: frontend/src/types/training.ts
/**
 * Types for ML Model Training, Cross Validation, and Diagnostic Evaluation.
 */

export type TaskType = 'classification' | 'regression';

export type ModelAlgorithm =
  | 'random_forest_classifier'
  | 'gradient_boosting_classifier'
  | 'logistic_regression'
  | 'decision_tree_classifier'
  | 'knn_classifier'
  | 'random_forest_regressor'
  | 'gradient_boosting_regressor'
  | 'linear_regression'
  | 'ridge_regression'
  | 'decision_tree_regressor'
  | 'knn_regressor';

export interface HyperparameterProperty {
  type: 'int' | 'float' | 'string';
  min?: number;
  max?: number;
  default: number | string;
}

export interface AlgorithmInfo {
  algorithm: string;
  name: string;
  task_type: TaskType;
  description: string;
  default_hyperparameters: Record<string, number | string>;
  hyperparameter_schema: Record<string, HyperparameterProperty>;
}

export interface CrossValidationConfig {
  n_splits: number;
  stratify: boolean;
  shuffle: boolean;
  random_state: number;
}

export interface TrainingJobCreate {
  project_id: string;
  dataset_id: string;
  pipeline_id?: string;
  target_column: string;
  task_type: TaskType;
  algorithm: string;
  hyperparameters: Record<string, number | string>;
  cv_config?: CrossValidationConfig;
  model_name?: string;
}

export interface EvaluationMetrics {
  // Classification
  accuracy?: number;
  precision?: number;
  recall?: number;
  f1_score?: number;
  roc_auc?: number;
  log_loss?: number;
  // Regression
  mse?: number;
  rmse?: number;
  mae?: number;
  r2_score?: number;
  explained_variance?: number;
}

export interface ConfusionMatrixData {
  labels: string[];
  matrix: number[][];
  normalized_matrix: number[][];
}

export interface CurvePoint {
  x: number;
  y: number;
  threshold?: number;
}

export interface CVFoldResult {
  fold: number;
  train_score: number;
  val_score: number;
  metrics: Record<string, number>;
}

export interface CVSummary {
  mean_val_score: number;
  std_val_score: number;
  folds: CVFoldResult[];
}

export interface FeatureImportanceItem {
  feature: string;
  importance: number;
  rank: number;
}

export interface ResidualPoint {
  actual: number;
  predicted: number;
  residual: number;
}

export interface TrainingJobDetailResponse {
  id: string;
  project_id: string;
  dataset_id: string;
  model_id?: string;
  status: string;
  algorithm: string;
  task_type: TaskType;
  target_column: string;
  hyperparameters: Record<string, unknown>;
  train_metrics: EvaluationMetrics;
  test_metrics: EvaluationMetrics;
  cv_summary?: CVSummary;
  feature_importances: FeatureImportanceItem[];
  confusion_matrix?: ConfusionMatrixData;
  roc_curve: CurvePoint[];
  residuals_sample: ResidualPoint[];
  training_duration_ms: number;
  artifact_path?: string;
  created_at: string;
}

export interface TrainingJobListItem {
  id: string;
  project_id: string;
  project_name: string;
  dataset_id: string;
  dataset_name: string;
  target_column: string;
  task_type: TaskType;
  status: 'PENDING' | 'RUNNING' | 'SUCCESS' | 'FAILED' | 'CANCELLED';
  error_message?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}
